### PR TITLE
Implement method type specialisation

### PR DESCRIPTION
**Pull request by AlexSikia**
This implements method type specialization.

@DarkDimius Please review (and of course let me know if I should do anything differently with the pull request!)
Thanks

### CODE REVIEW

File: bench/src/test/test.iml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (DarkDimius) [inline bench/src/test/test.iml]: This shouldn't be part of PR.

**Comment by DarkDimius** [comment]
Alex, a lot of your commits do reformatting of already existing code for not good reason(I guess your IDE has different formatting options), could you please update your commits to not do this, as this affects git history.


Also, you have committed files that are specific to your IDE(I guess that's Intellij) that shouldn't be part of PR.

Neither should there be anything in out/ folder.

**Comment by DarkDimius** [comment]
Junit failed in GenBCode, as output directory does not exist..
`out/.keep` should be kept.

**Comment by DarkDimius** [comment]
Great to see all the tests pass :+1: 

PS: this PR only contains method specialization, it does not handle classes or fields.

File: src/dotty/tools/dotc/transform/TypeSpecializer.scala [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+  private val newSymbolMap: mutable.HashMap[Symbol, mutable.HashMap[List[Type], Symbols.Symbol]] = mutable.HashMap.empty
+
+  def allowedToSpecialize(sym: Symbol, numOfTypes: Int)(implicit ctx: Context): Boolean = {
+    numOfTypes <= maxTparamsToSpecialize &&
+      numOfTypes > 0 &&
+      sym.name != nme.asInstanceOf_ &&
+      sym.name != nme.isInstanceOf_ &&
+      !(sym is Flags.JavaDefined) &&
+      !sym.isConstructor &&
+      !sym.name.toString.contains("Function2")
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: why?

**Comment by VladUreche** [comment]
How did you guys fix the transitivity issue?

**Comment by DarkDimius** [comment]
As a last transformation step, in all places where expected type propagation starts, [casts are introduced](https://github.com/AlexSikia/dotty/commit/bdfb1c1233f80eca7bce755ad2ea4bb100d1087d#diff-a979e1f98bfa8b4c0d7cba7a8f64f8f6R193).

This isn't a solution, this is a hacky-workaround. I do not see a good solution besides #592.

File: src/dotty/tools/dotc/core/Definitions.scala [hunk 3]
```diff
@@ -326,6 +326,7 @@ class Definitions {
   lazy val TransientAnnot                = ctx.requiredClass("scala.transient")
   lazy val NativeAnnot                   = ctx.requiredClass("scala.native")
   lazy val ScalaStrictFPAnnot            = ctx.requiredClass("scala.annotation.strictfp")
+  lazy val specializedAnnot              = ctx.requiredClass("scala.specialized")
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/Definitions.scala]: capital S

File: tests/pos/specialization/bounds_specialization.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
+object bounds_specialization {
+  /*class Foo[@specialized K] {
```
Comment (DarkDimius) [inline tests/pos/specialization/bounds_specialization.scala]: why is this test commented?

File: tests/pos/specialization/nothing_specialization.scala [hunk 5]
```diff
@@ -0,0 +1,8 @@
+object nothing_specialization {
+  def ret_nothing[@specialized(Char) T] = {
+    //val a: List[T] = List[Nothing]()
```
Comment (DarkDimius) [inline tests/pos/specialization/nothing_specialization.scala]: why is it commented?

File: src/dotty/tools/dotc/transform/PreSpecializer.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * This phase retrieves all `@specialized` anotations before they are thrown away,
+ * and stores them for the `TypeSpecializer` phase.
+ */
+class PreSpecializer extends MiniPhaseTransform {
+
+  override def phaseName: String = "prespecialize"
+
+  private final def primitiveCompanionToPrimitive(companion: Type)(implicit ctx: Context) = {
+    if (companion.asInstanceOf[TermRef].name.toString == "AnyRef") { // Handles `@specialized(AnyRef)` cases
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/PreSpecializer.scala]: why based on names, not on symbols?

**Comment by DarkDimius** [comment]
```scala
trait Spec {
 def plus[@specialized T](a: T, b:T)(ev: Numeric[T]): T = plus(b, a)(ev)
}
```

Generates specialized versions, but does not dispatch to them in recursive call.

**Comment by VladUreche** [comment]
@DarkDimius @AlexSikia doesn't casting introduce boxes?

**Comment by DarkDimius** [comment]
@VladUreche the problems we've been seeing here mostly have to do with type arguments of enclosing scope, that are bounded by specialized types defined in inner scope. This happens infrequently in real code, and when it happens, it's mostly in higher kinded types, ie casts will be erased.

Examples can be found here: https://github.com/AlexSikia/dotty/blob/add/Type-specialisation/tests/pos/specialization/this_specialization.scala

But yes, it could potentially introduce boxing, and this will could some calls from being dispatched to specialized versions.

File: src/dotty/tools/dotc/transform/TypeSpecializer.scala [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        }).flatten
+      }
+      else {
+        generateSpecializedSymbols(instantiations.reverse, names.reverse, poly, decl)
+      }
+    }
+    def generateSpecializedSymbols(instantiations: List[Type], names: List[String], poly: PolyType, decl: Symbol)
+                                  (implicit ctx: Context): List[Symbol] = {
+      val newSym =
+        ctx.newSymbol(decl.owner, (decl.name + "$mc" + names.mkString + "$sp").toTermName,
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: If I understand the name mangling of specialization correctly, this should be `(decl.name + "$m" + names.mkString + "c$sp")`.
 `m` means method and `c` means class, for example if there's a specialized Int method in a specialized Int class, the signature will contain `$mIcI$sp`.

File: src/dotty/tools/dotc/transform/TypeSpecializer.scala [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    Map(defn.ByteType -> "B",
+      defn.BooleanType -> "Z",
+      defn.ShortType -> "S",
+      defn.IntType -> "I",
+      defn.LongType -> "J",
+      defn.FloatType -> "F",
+      defn.DoubleType -> "D",
+      defn.CharType -> "C",
+      defn.UnitType -> "V",
+      defn.AnyRefType -> "L")
```
Comment (retronym) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: This seems inefficient, the calling code `specialisedTypeToSuffix.apply(tp)` created the entire map and then throws it away.

Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: This function is simply not needed. There's already a `specializedFor` in NameOps: https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/NameOps.scala#L239

Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: Thought this method currently does not handle method type arguments. This functionality should be added.

Comment (AlexSikia) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: Thanks for pointing this out! I'll take care of changing this when I reach Geneva later today.

File: src/dotty/tools/dotc/transform/TypeSpecializer.scala (hunk 7 above)
Comment (AlexSikia) [inline src/dotty/tools/dotc/transform/TypeSpecializer.scala]: Ah, good point; thanks for finding that.

**Comment by DarkDimius** [comment]
needs to be rebased over current master.

**Comment by DarkDimius** [comment]
@AlexSikia, This PR has been around for quite some time. Let's do a push and get this in?

**Comment by odersky** [comment]
I think it would be better to keep this separate until we have tackled the larger problem of class specialization.

**Comment by smarter** [comment]
Commit https://github.com/AlexSikia/dotty/commit/07e9ae12d574fa0bbb07012cd78611973e2afeb7 should be removed from this pull request, currently it's just an accidental revert of https://github.com/lampepfl/dotty/commit/10ea7125998b0246e899706f407cc57719f37086 .

**Comment by odersky** [comment]
Superseded by new work

**Pull request closed without merging**